src/algebra/category/CommRing/basic.lean

```diff
@@ -18,103 +18,123 @@ We introduce the bundled categories:
 * `CommSemiRing`
 * `CommRing`
 along with the relevant forgetful functors between them.
+
+## Implementation notes
+
+We make SemiRing (and the other categories) locally reducible in order
+to define its instances. This is because writing, for example,
+
+  instance : concrete_category SemiRing := by { delta SemiRing, apply_instance }
+
+results in an instance of the form `id (bundled_hom.concrete_category _)`
+and this `id`, not being [reducible], prevents a later instance search
+(once SemiRing is no longer reducible) from seeing that the morphisms of
+SemiRing are really semiring morphisms (`→+*`), and therefore have a coercion
+to functions, for example. It's especially important that the `has_coe_to_sort`
+instance not contain an extra `id` as we want the `semiring ↥R` instance to
+also apply to `semiring R.α` (it seems to be impractical to guarantee that
+we always access `R.α` through the coercion rather than directly).
+
+TODO: Probably @[derive] should be able to create instances of the
+required form (without `id`), and then we could use that instead of
+this obscure `local attribute [reducible]` method.
+
 -/
 
 universes u v
 
 open category_theory
 
 /-- The category of semirings. -/
-@[reducible] def SemiRing : Type (u+1) := bundled semiring
+def SemiRing : Type (u+1) := bundled semiring
 
 namespace SemiRing
 
 /-- Construct a bundled SemiRing from the underlying type and typeclass. -/
 def of (R : Type u) [semiring R] : SemiRing := bundled.of R
 
+local attribute [reducible] SemiRing
+
+instance : has_coe_to_sort SemiRing := infer_instance
+
 instance (R : SemiRing) : semiring R := R.str
 
 instance bundled_hom : bundled_hom @ring_hom :=
 ⟨@ring_hom.to_fun, @ring_hom.id, @ring_hom.comp, @ring_hom.coe_inj⟩
 
-instance has_forget_to_Mon : has_forget₂ SemiRing.{u} Mon.{u} :=
+instance : concrete_category SemiRing := infer_instance
+
+instance has_forget_to_Mon : has_forget₂ SemiRing Mon :=
 bundled_hom.mk_has_forget₂ @semiring.to_monoid (λ R₁ R₂ f, f.to_monoid_hom) (λ _ _ _, rfl)
 
 end SemiRing
 
 /-- The category of rings. -/
-@[reducible] def Ring : Type (u+1) := bundled ring
+def Ring : Type (u+1) := induced_category SemiRing (bundled.map @ring.to_semiring)
 
 namespace Ring
 
-instance (R : Ring) : ring R := R.str
-
 /-- Construct a bundled Ring from the underlying type and typeclass. -/
 def of (R : Type u) [ring R] : Ring := bundled.of R
 
-instance bundled_hom : bundled_hom _ :=
-SemiRing.bundled_hom.full_subcategory @ring.to_semiring
+local attribute [reducible] Ring
+
+instance : has_coe_to_sort Ring := infer_instance
 
-instance has_forget_to_SemiRing : has_forget₂ Ring.{u} SemiRing.{u} :=
-SemiRing.bundled_hom.full_subcategory_has_forget₂ _
+instance (R : Ring) : ring R := R.str
+
+instance : concrete_category Ring := infer_instance
+
+instance has_forget_to_SemiRing : has_forget₂ Ring SemiRing := infer_instance
```

Comment:
Do we also want a forgetful functor to `AddGroup`?

Reply:
And this (to `AddCommGroup`).